cumulusci/tasks/bulkdata.py

```diff
@@ -1,63 +1,119 @@
-from cumulusci.tasks.salesforce import BaseSalesforceApiTask
+from future import standard_library
 
+standard_library.install_aliases()
+from builtins import zip
+from contextlib import contextmanager
+from urllib.parse import urljoin
 import csv
-import time
-import hiyapyco
-import xml.etree.ElementTree as ET
-
 import datetime
-import requests
+import io
+import itertools
+import time
 import tempfile
-import unicodecsv
-
-from collections import OrderedDict
-
-from salesforce_bulk import CsvDictsAdapter
+import xml.etree.ElementTree as ET
 
+from salesforce_bulk.util import IteratorBytesIO
 from sqlalchemy.ext.automap import automap_base
+from sqlalchemy.orm import aliased
 from sqlalchemy.orm import create_session
 from sqlalchemy.orm import mapper
 from sqlalchemy.orm import Session
 from sqlalchemy import create_engine
 from sqlalchemy import Column
-from sqlalchemy import Integer
 from sqlalchemy import MetaData
 from sqlalchemy import Table
 from sqlalchemy import Unicode
 from sqlalchemy import text
 from sqlalchemy import types
 from sqlalchemy import event
+import hiyapyco
+import requests
+import unicodecsv
+
+from cumulusci.tasks.salesforce import BaseSalesforceApiTask
 
 # TODO: UserID Catcher
 # TODO: Dater
 
-# Create a custom sqlalchemy field type for sqlite datetime fields which are stored as integer of epoch time
-class EpochType(types.TypeDecorator):
-    impl = types.Integer
-
-    epoch = datetime.datetime(1970, 1, 1, 0, 0, 0)
-
-    def process_bind_param(self, value, dialect):
-        return int((value - self.epoch).total_seconds()) * 1000
 
-    def process_result_value(self, value, dialect):
-        return self.epoch + datetime.timedelta(seconds=value // 1000)
-
-
-# Listen for sqlalchemy column_reflect event and map datetime fields to EpochType
-@event.listens_for(Table, "column_reflect")
-def setup_epoch(inspector, table, column_info):
-    if isinstance(column_info["type"], types.DateTime):
-        column_info["type"] = EpochType()
+class BulkJobTaskMixin(object):
+    def _job_state_from_batches(self, job_id):
+        uri = urljoin(self.bulk.endpoint + "/", "job/{0}/batch".format(job_id))
+        response = requests.get(uri, headers=self.bulk.headers())
+        return self._parse_job_state(response.content)
+
+    def _parse_job_state(self, xml):
+        tree = ET.fromstring(xml)
+        completed = 0
+        pending = 0
+        failed = 0
+        for el in tree.iterfind(".//{%s}state" % self.bulk.jobNS):
+            state = el.text
+            if state == "Not Processed":
+                return "Aborted"
+            elif state == "Failed":
+                failed += 1
+            elif state == "Completed":
+                completed += 1
+            else:  # Queued, InProgress
+                pending += 1
+        if pending:
+            return "InProgress"
+        elif failed:
+            return "Failed"
+        else:
+            return "Completed"
+
+    def _wait_for_job(self, job_id):
```

Comment:
should this use the _poller from basetask? if its a "yes but it doesnt do X" can we capture X in an issue? we know we're gonna need to update the poller a bit so if we can capture requests thats nice

Reply:
Oh, probably -- well, I'd like to turn the poller into a separate utility function or class.